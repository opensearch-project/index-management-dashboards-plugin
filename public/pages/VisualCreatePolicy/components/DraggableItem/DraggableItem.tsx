/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { EuiFlexGroup, EuiFlexItem, EuiDraggable, EuiPanel, EuiIcon, EuiButtonIcon, EuiToolTip } from "@elastic/eui";

interface DraggableItemProps {
  content: string | JSX.Element | null;
  id: string;
  idx: number;
  isLast: boolean;
  onClickDelete: () => void;
  onClickEdit: () => void;
  draggableType: "action" | "transition";
}

const DraggableItem = ({ content, id, idx, isLast, onClickDelete, onClickEdit, draggableType }: DraggableItemProps) => (
  <EuiDraggable style={{ padding: `0px 0px ${isLast ? "0px" : "10px"} 0px` }} key={id} index={idx} draggableId={id} customDragHandle={true}>
    {(provided) => (
      <EuiPanel className="custom" paddingSize="m">
        <EuiFlexGroup alignItems="center">
          <EuiFlexItem grow={false}>
            <div {...provided.dragHandleProps} aria-label="Drag Handle">
              <EuiIcon type="grab" />
            </div>
          </EuiFlexItem>
          <EuiFlexItem>{content}</EuiFlexItem>
          <EuiFlexItem grow={false}>
            <EuiToolTip position="top" content={<p>Delete {draggableType}</p>}>
              <EuiButtonIcon
                iconType="trash"
                aria-label="Delete"
                color="danger"
                onClick={onClickDelete}
                data-test-subj={`draggable-item-delete-button-${id}`}
              />
            </EuiToolTip>
          </EuiFlexItem>
          <EuiFlexItem grow={false}>
            <EuiToolTip position="top" content={<p>Edit {draggableType}</p>}>
              <EuiButtonIcon
                iconType="pencil"
                aria-label="Edit"
                color="primary"
                onClick={onClickEdit}
                data-test-subj={`draggable-item-edit-button-${id}`}
              />
            </EuiToolTip>
          </EuiFlexItem>
        </EuiFlexGroup>
      </EuiPanel>
    )}
  </EuiDraggable>
);

export default DraggableItem;
