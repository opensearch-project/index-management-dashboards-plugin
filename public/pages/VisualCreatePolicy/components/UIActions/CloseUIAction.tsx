/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { CloseAction, UIAction } from "../../../../../models/interfaces";
import { makeId } from "../../../../utils/helpers";
import { ActionType } from "../../utils/constants";

export default class CloseUIAction implements UIAction<CloseAction> {
  id: string;
  action: CloseAction;
  type = ActionType.Close;

  constructor(action: CloseAction, id: string = makeId()) {
    this.action = action;
    this.id = id;
  }

  content = () => `Close`;

  clone = (action: CloseAction) => new CloseUIAction(action, this.id);

  isValid = () => true;

  render = (action: UIAction<CloseAction>, onChangeAction: (action: UIAction<CloseAction>) => void) => {
    return <div />;
  };

  toAction = () => this.action;
}
