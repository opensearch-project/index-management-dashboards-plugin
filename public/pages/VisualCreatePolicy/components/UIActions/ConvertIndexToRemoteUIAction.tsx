/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { ChangeEvent } from "react";
import { EuiCompressedFormRow, EuiCompressedFieldText, EuiSpacer } from "@elastic/eui";
import { ConvertIndexToRemoteAction, UIAction } from "../../../../../models/interfaces";
import { makeId } from "../../../../utils/helpers";
import { ActionType } from "../../utils/constants";
import EuiFormCustomLabel from "../EuiFormCustomLabel";

export default class ConvertIndexToRemoteUIAction implements UIAction<ConvertIndexToRemoteAction> {
  id: string;
  action: ConvertIndexToRemoteAction;
  type = ActionType.ConvertIndexToRemote;

  constructor(action: ConvertIndexToRemoteAction, id: string = makeId()) {
    this.action = action;
    this.id = id;
  }

  content = () => `Convert Index To Remote`;

  clone = (action: ConvertIndexToRemoteAction) => new ConvertIndexToRemoteUIAction(action, this.id);

  isValid = () => {
    return !!this.action.convert_index_to_remote.snapshot && !!this.action.convert_index_to_remote.repository;
  };

  render = (action: UIAction<ConvertIndexToRemoteAction>, onChangeAction: (action: UIAction<ConvertIndexToRemoteAction>) => void) => {
    return (
      <>
        <EuiFormCustomLabel
          title="Repository"
          helpText="The repository name that you register through the native snapshot API operations."
          isInvalid={!this.isValid()}
        />
        <EuiCompressedFormRow fullWidth isInvalid={!this.isValid()} error={null}>
          <EuiCompressedFieldText
            fullWidth
            value={(action.action as ConvertIndexToRemoteAction).convert_index_to_remote.repository}
            onChange={(e: ChangeEvent<HTMLInputElement>) => {
              const repository = e.target.value;
              onChangeAction(
                this.clone({
                  convert_index_to_remote: {
                    ...action.action.convert_index_to_remote,
                    repository,
                  },
                })
              );
            }}
            data-test-subj="action-render-convert-index-to-remote-repository"
          />
        </EuiCompressedFormRow>
        <EuiSpacer size="s" />
        <EuiFormCustomLabel title="Snapshot" helpText="The name of the snapshot." isInvalid={!this.isValid()} />
        <EuiCompressedFormRow fullWidth isInvalid={!this.isValid()} error={null}>
          <EuiCompressedFieldText
            fullWidth
            value={(action.action as ConvertIndexToRemoteAction).convert_index_to_remote.snapshot}
            onChange={(e: ChangeEvent<HTMLInputElement>) => {
              const snapshot = e.target.value;
              onChangeAction(
                this.clone({
                  convert_index_to_remote: {
                    ...action.action.convert_index_to_remote,
                    snapshot,
                  },
                })
              );
            }}
            data-test-subj="action-render-convert-index-to-remote-snapshot"
          />
        </EuiCompressedFormRow>
      </>
    );
  };

  toAction = () => this.action;
}
