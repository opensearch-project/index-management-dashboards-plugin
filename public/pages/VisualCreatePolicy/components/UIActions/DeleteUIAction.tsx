/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { DeleteAction, UIAction } from "../../../../../models/interfaces";
import { makeId } from "../../../../utils/helpers";
import { ActionType } from "../../utils/constants";

export default class DeleteUIAction implements UIAction<DeleteAction> {
  id: string;
  action: DeleteAction;
  type = ActionType.Delete;

  constructor(action: DeleteAction, id: string = makeId()) {
    this.action = action;
    this.id = id;
  }

  content = () => `Delete`;

  clone = (action: DeleteAction) => new DeleteUIAction(action, this.id);

  isValid = () => true;

  render = (action: UIAction<DeleteAction>, onChangeAction: (action: UIAction<DeleteAction>) => void) => {
    return <div />;
  };

  toAction = () => this.action;
}
