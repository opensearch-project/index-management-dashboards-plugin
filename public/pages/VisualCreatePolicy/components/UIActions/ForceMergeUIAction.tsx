/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { ChangeEvent } from "react";
import { EuiFormRow, EuiFieldNumber } from "@elastic/eui";
import EuiFormCustomLabel from "../EuiFormCustomLabel";
import { ForceMergeAction, UIAction } from "../../../../../models/interfaces";
import { makeId } from "../../../../utils/helpers";
import { ActionType } from "../../utils/constants";

export default class ForceMergeUIAction implements UIAction<ForceMergeAction> {
  id: string;
  action: ForceMergeAction;
  type = ActionType.ForceMerge;

  constructor(action: ForceMergeAction, id: string = makeId()) {
    this.action = action;
    this.id = id;
  }

  content = () => `Force merge to ${this.action.force_merge.max_num_segments} segments`;

  clone = (action: ForceMergeAction) => new ForceMergeUIAction(action, this.id);

  isValid = () => {
    const segments = this.action.force_merge.max_num_segments;
    return !!segments && segments > 0;
  };

  render = (action: UIAction<ForceMergeAction>, onChangeAction: (action: UIAction<ForceMergeAction>) => void) => {
    const segments = action.action.force_merge.max_num_segments;
    return (
      <>
        <EuiFormCustomLabel title="Max num segments" helpText="The number of segments to merge to." isInvalid={!this.isValid()} />
        <EuiFormRow fullWidth isInvalid={!this.isValid()} error={null}>
          <EuiFieldNumber
            fullWidth
            value={typeof segments === "undefined" ? "" : segments}
            onChange={(e: ChangeEvent<HTMLInputElement>) => {
              const maxNumSegments = e.target.valueAsNumber;
              const forceMerge = { max_num_segments: maxNumSegments };
              if (isNaN(maxNumSegments)) delete forceMerge.max_num_segments;
              onChangeAction(this.clone({ force_merge: forceMerge }));
            }}
            data-test-subj="action-render-force-merge"
          />
        </EuiFormRow>
      </>
    );
  };

  toAction = () => this.action;
}
