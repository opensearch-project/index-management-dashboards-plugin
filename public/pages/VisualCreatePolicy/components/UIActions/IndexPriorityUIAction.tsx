/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { ChangeEvent } from "react";
import { EuiFormRow, EuiFieldNumber } from "@elastic/eui";
import EuiFormCustomLabel from "../EuiFormCustomLabel";
import { IndexPriorityAction, UIAction } from "../../../../../models/interfaces";
import { makeId } from "../../../../utils/helpers";
import { ActionType } from "../../utils/constants";

export default class IndexPriorityUIAction implements UIAction<IndexPriorityAction> {
  id: string;
  action: IndexPriorityAction;
  type = ActionType.IndexPriority;

  constructor(action: IndexPriorityAction, id: string = makeId()) {
    this.action = action;
    this.id = id;
  }

  content = () => `Set index priority to ${this.action.index_priority.priority}`;

  clone = (action: IndexPriorityAction) => new IndexPriorityUIAction(action, this.id);

  isValid = () => {
    const priority = this.action.index_priority.priority;
    return typeof priority !== "undefined" && priority >= 0;
  };

  render = (action: UIAction<IndexPriorityAction>, onChangeAction: (action: UIAction<IndexPriorityAction>) => void) => {
    const priority = action.action.index_priority.priority;
    return (
      <>
        <EuiFormCustomLabel
          title="Priority"
          helpText="Higher priority indices are recovered first when possible."
          isInvalid={!this.isValid()}
        />
        <EuiFormRow fullWidth isInvalid={!this.isValid()} error={null}>
          <EuiFieldNumber
            fullWidth
            value={typeof priority === "undefined" ? "" : priority}
            onChange={(e: ChangeEvent<HTMLInputElement>) => {
              // eslint-disable-next-line no-shadow
              const priority = e.target.valueAsNumber;
              const indexPriority = { priority };
              if (isNaN(priority)) delete indexPriority.priority;
              onChangeAction(this.clone({ index_priority: indexPriority }));
            }}
            data-test-subj="action-render-index-priority"
          />
        </EuiFormRow>
      </>
    );
  };

  toAction = () => this.action;
}
