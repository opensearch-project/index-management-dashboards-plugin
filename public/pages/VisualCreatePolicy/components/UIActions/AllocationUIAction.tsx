/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { EuiFormRow, EuiCodeEditor } from "@elastic/eui";
import { AllocationAction, UIAction } from "../../../../../models/interfaces";
import { makeId } from "../../../../utils/helpers";
import { ActionType } from "../../utils/constants";
import { DarkModeConsumer } from "../../../../components/DarkMode";

export default class AllocationUIAction implements UIAction<AllocationAction> {
  id: string;
  action: AllocationAction;
  type = ActionType.Allocation;

  constructor(action: AllocationAction, id: string = makeId()) {
    this.action = action;
    this.id = id;
  }

  content = () => `Allocation`;

  clone = (action: AllocationAction = this.action) => new AllocationUIAction(action, this.id);

  isValid = () => {
    try {
      JSON.parse(this.getActionJsonString(this.action));
      return true;
    } catch (err) {
      return false;
    }
  };

  getActionJsonString = (action: AllocationAction) => {
    const allocation = action.allocation;
    return allocation.hasOwnProperty("jsonString") ? allocation.jsonString : JSON.stringify(allocation, null, 4);
  };

  getActionJson = (action: AllocationAction) => {
    const allocation = action.allocation;
    return allocation.hasOwnProperty("jsonString") ? JSON.parse(allocation.jsonString) : allocation;
  };

  render = (action: UIAction<AllocationAction>, onChangeAction: (action: UIAction<AllocationAction>) => void) => {
    return (
      <EuiFormRow fullWidth isInvalid={!this.isValid()} error={null} style={{ maxWidth: "100%" }}>
        <DarkModeConsumer>
          {(isDarkMode) => (
            <EuiCodeEditor
              mode="json"
              theme={isDarkMode ? "sense-dark" : "github"}
              width="100%"
              value={this.getActionJsonString(action.action)}
              onChange={(str) => {
                onChangeAction(
                  this.clone({
                    ...action.action,
                    allocation: { jsonString: str },
                  })
                );
              }}
              setOptions={{ fontSize: "14px" }}
              aria-label="Code Editor"
            />
          )}
        </DarkModeConsumer>
      </EuiFormRow>
    );
  };

  toAction = () => ({
    ...this.action,
    allocation: this.getActionJson(this.action),
  });
}
