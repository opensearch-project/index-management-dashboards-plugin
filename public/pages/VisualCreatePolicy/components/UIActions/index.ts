/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import AliasUIAction from "./AliasUIAction/AliasUIAction";
import AllocationUIAction from "./AllocationUIAction";
import CloseUIAction from "./CloseUIAction";
import DeleteUIAction from "./DeleteUIAction";
import ForceMergeUIAction from "./ForceMergeUIAction";
import IndexPriorityUIAction from "./IndexPriorityUIAction";
import NotificationUIAction from "./NotificationUIAction";
import OpenUIAction from "./OpenUIAction";
import ReadOnlyUIAction from "./ReadOnlyUIAction";
import ReadWriteUIAction from "./ReadWriteUIAction";
import ReplicaCountUIAction from "./ReplicaCountUIAction";
import RolloverUIAction from "./RolloverUIAction";
import RollupUIAction from "./RollupUIAction";
import ShrinkUIAction from "./ShrinkUIAction";
import SnapshotUIAction from "./SnapshotUIAction";

export {
  AliasUIAction,
  AllocationUIAction,
  CloseUIAction,
  DeleteUIAction,
  ForceMergeUIAction,
  IndexPriorityUIAction,
  NotificationUIAction,
  OpenUIAction,
  ReadOnlyUIAction,
  ReadWriteUIAction,
  ReplicaCountUIAction,
  RolloverUIAction,
  RollupUIAction,
  ShrinkUIAction,
  SnapshotUIAction,
};
