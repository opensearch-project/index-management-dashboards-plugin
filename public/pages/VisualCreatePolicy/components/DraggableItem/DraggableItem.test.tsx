/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render } from "@testing-library/react";
import { EuiDragDropContext, EuiDroppable } from "@elastic/eui";
import { fireEvent } from "@testing-library/dom";
import DraggableItem from "./DraggableItem";
import { DEFAULT_ROLLOVER } from "../../utils/constants";
import { RolloverUIAction } from "../UIActions";
import { UIAction } from "../../../../../models/interfaces";

describe("<DraggableItem /> spec", () => {
  it("renders the component", () => {
    const action: UIAction<any> = new RolloverUIAction(DEFAULT_ROLLOVER);
    const content = action.content();
    const { container } = render(
      <EuiDragDropContext onDragEnd={() => {}}>
        <EuiDroppable droppableId="STATE_ACTIONS_DROPPABLE_AREA">
          <DraggableItem
            content={content}
            id={action.id}
            idx={0}
            isLast={true}
            onClickDelete={() => {}}
            onClickEdit={() => {}}
            draggableType="action"
          />
        </EuiDroppable>
      </EuiDragDropContext>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("calls onclickdeleteaction when clicking delete button", () => {
    const action: UIAction<any> = new RolloverUIAction(DEFAULT_ROLLOVER);
    const onClickDelete = jest.fn();
    const content = action.content();
    const { getByTestId } = render(
      <EuiDragDropContext onDragEnd={() => {}}>
        <EuiDroppable droppableId="STATE_ACTIONS_DROPPABLE_AREA">
          <DraggableItem
            content={content}
            id={action.id}
            idx={0}
            isLast={true}
            onClickDelete={onClickDelete}
            onClickEdit={() => {}}
            draggableType="action"
          />
        </EuiDroppable>
      </EuiDragDropContext>
    );

    fireEvent.click(getByTestId("draggable-item-delete-button-" + action.id));
    expect(onClickDelete).toHaveBeenCalledTimes(1);
  });

  it("calls onclickeditaction when clicking edit button", () => {
    const action: UIAction<any> = new RolloverUIAction(DEFAULT_ROLLOVER);
    const onClickEdit = jest.fn();
    const content = action.content();
    const { getByTestId } = render(
      <EuiDragDropContext onDragEnd={() => {}}>
        <EuiDroppable droppableId="STATE_ACTIONS_DROPPABLE_AREA">
          <DraggableItem
            content={content}
            id={action.id}
            idx={0}
            isLast={true}
            onClickDelete={() => {}}
            onClickEdit={onClickEdit}
            draggableType="action"
          />
        </EuiDroppable>
      </EuiDragDropContext>
    );

    fireEvent.click(getByTestId("draggable-item-edit-button-" + action.id));
    expect(onClickEdit).toHaveBeenCalledTimes(1);
  });
});
