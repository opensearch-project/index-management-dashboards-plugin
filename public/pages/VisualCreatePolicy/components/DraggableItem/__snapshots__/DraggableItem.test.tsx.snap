// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DraggableItem /> spec renders the component 1`] = `
<div
  class="euiDroppable euiDroppable--noGrow"
  data-rbd-droppable-context-id="0"
  data-rbd-droppable-id="STATE_ACTIONS_DROPPABLE_AREA"
  data-test-subj="droppable"
>
  <div
    class="euiDraggable euiDraggable--hasCustomDragHandle"
    data-rbd-draggable-context-id="0"
    data-rbd-draggable-id="some_html_id"
    data-test-subj="draggable"
    style="padding: 0px 0px 0px 0px;"
  >
    <div
      class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow custom euiDraggable__item euiDraggable__item--hasCustomDragHandle"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <div
            aria-describedby="rbd-hidden-text-0-hidden-text-0"
            aria-label="Drag Handle"
            data-rbd-drag-handle-context-id="0"
            data-rbd-drag-handle-draggable-id="some_html_id"
            draggable="false"
            role="button"
            tabindex="0"
          >
            <svg
              aria-hidden="true"
              class="euiIcon euiIcon--medium euiIcon-isLoading"
              focusable="false"
              height="16"
              role="img"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            />
          </div>
        </div>
        <div
          class="euiFlexItem"
        >
          Rollover
        </div>
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <span
            class="euiToolTipAnchor"
          >
            <button
              aria-label="Delete"
              class="euiButtonIcon euiButtonIcon--danger euiButtonIcon--empty euiButtonIcon--xSmall"
              data-test-subj="draggable-item-delete-button-some_html_id"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonIcon__icon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              />
            </button>
          </span>
        </div>
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <span
            class="euiToolTipAnchor"
          >
            <button
              aria-label="Edit"
              class="euiButtonIcon euiButtonIcon--primary euiButtonIcon--empty euiButtonIcon--xSmall"
              data-test-subj="draggable-item-edit-button-some_html_id"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonIcon__icon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              />
            </button>
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiDroppable__placeholder"
  />
</div>
`;
